숫자야구 [Step 1] Rilla, Chirsty

juun97 opened the pull request:
@Gundy93 
건디! 첫 PR 입니다! 잘부탁드려요 ㅎㅎ 👏🏻👏🏻

## iOS-Number-Baseball [Step1] By Rilla, Christy
---

## 고민했던 부분💡

#### 중복되지 않은 3개의 숫자
>1에서 9까지의 숫자중 중복되지 않은 숫자를 어떻게 3개만 뽑아 올까 고민이 되었어요!
만약 뽑았을때 중복의 갯수가 한개가 될 수도 있고 두개가 될 수도 있었기에 평범한 배열이나 조건문으로는 쉽게 해결하지 못할것 같았어요! 그래서 열심히 구글링 한 결과 중복된 값을 허용하지 않은 배열 Set을 발견했고 이를 사용해 Count가 3이 될때까지 뽑아 중복되지 않은 수를 3개만 뽑아올 수 있었습니다!

#### FlowChart 작성법
>기존에 FlowChart작성을 해본적이 없어서 어디까지 그림에 표현해야 하는지에 대한 감이 잘 잡히질 않았어요!
그래서 크리스티랑 열심히 머리를 맞댄 결과 최소한의 흐름도만 작성하자 라는 결과에 도출 했고 Step1 에서 요구하는 최소 기능만을 담아서 커밋 했습니다!

--- 

## 조언이 필요한 부분 ❓

#### 변수의 추가적인 생성

'''
 
    
    while temp > 0 {
        let randomNumbers = makeRandomNumbers()
        result = checkNumbers(numbers: randomNumbers)
        let strike = result[0]
        let ball = result[1]
        
        if strike == 3 {
            print("정답 입니다!")
            return
        }
        
        temp -= 1
        
        print("임의의 수 : \(randomNumbers[0]), \(randomNumbers[1]), \(randomNumbers[2])")
        print("\(strike) 스트라이크, \(ball) 볼")
        print("남은 기회 : \(temp)")
    }
    
    print("컴퓨터의 승리입니다.")
    


'''
이 부분에서 result[0] 과 result[1] 을 하단 print에서 바로 사용할 수 있으나 변수를 할당해서 사용했습니다!
그런데 프로젝트의 규모가 커지게 된다면 생성하는 변수도 훨씬 많아질텐데 어떤게 더 나을까요??
변수를 계속 새로 할당을 해야할까요 아니면 바로 써서 코드의 사이즈를 줄이는게 나을까요?

christy-hs-lee committed:
chore: 테스트 커밋

christy-hs-lee committed:
chore: 테스트 커밋

christy-hs-lee committed:
docs: readme 수정

christy-hs-lee committed:
chore: test 커밋

christy-hs-lee committed:
Merge branch 'main' of https://github.com/christy-hs-lee/ios-number-baseball into main

juun97 committed:
Chore: Test

Rilla's Test

christy-hs-lee committed:
chore: test commit

juun97 committed:
feat : 랜덤 상수 추출 메서드

중복되지 않은 1에서 9까지의 3가지 숫자를 추출해서 배열로 리턴하는 메서드 구현

christy-hs-lee committed:
fix : makeRandomNumbers 메서드 버그 수정

christy-hs-lee committed:
feat : checkNumbers 메소드 생성

기존값과 임의의 수를 비교해 strike와 ball을 확인하는 메소드

christy-hs-lee committed:
feat : playGame 메서드 생성

비교 후 strike와 ball의 갯수를 출력하고 정답을 확인하는 메서드

juun97 committed:
fix : playGame() 메서드 수정

부분 완성된 playGame() 메서드를 완성 시켰습니다.

juun97 committed:
Merge pull request #1 from christy-hs-lee/step_1

ios-number-baseball [Step 1] Rilla, Christy

christy-hs-lee committed:
feat : menuPrint 메서드 생성

프로그램 실행 함수이며 사용자에게 입력값을 받아 게임실행 혹은 게임종료를 실행하는 메서드

juun97 committed:
feat,fix : checkAvailability 메서드 생성 및 playGame 메서드 수정

checkAvailability() : 입력값을 전달인자로 받아 유효성 검사를 하는 메서드

playGame() : 기존 임의의 숫자를 받던 메서드를 사용자 입력을 받는 메서드로 수정

christy-hs-lee committed:
docs : step2 flow chart 추가

christy-hs-lee committed:
fix : 리드미 수정

step2 flow chart 수정

christy-hs-lee committed:
fix : 오타 수정

오타 수정

christy-hs-lee committed:
fix : step1 코드

step1 코드

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,66 @@
 
 import Foundation
```
현재 프로젝트에서, 이 파일에서 `import Foundation`이 필요한 코드일까요?
`import`에 대해서 알아보시는 것도 좋을 것 같습니다 😃

christy-hs-lee replied:
Foundation은 라이브러리중 하나로 우리가 아무렇지 않게 사용했던 기본 Type 과 배열의 기능들이 Foundation에서 제공해주고 있었다는걸 알게 됐어요! 따라서 변수 등 다양한 처리를 위해 이 파일에서 필수적인 코드라고 생각합니다 😄

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,66 @@
 import Foundation
 
-print("Hello, World!")
+let numbers = Array(1...9)
```
`numbers`는 현재 전역 변수로 선언되어 있는데, 전역 변수와 지역 변수의 차이에 대해 한번 정리해보셔도 좋을 것 같아요
정리한 사항을 답글로 적어주시면 좋겠습니다 💪

juun97 replied:
지역 변수는 코드 블럭안에서 생성된 변수로 stack 메모리에서 관리하게 됩니다! 이 변수는 해당 블록에서만 사용이 가능하며 변수가 속해있는 코드블럭이 끝나면 stack 메모리에서 해제가 되며 삭제가 됩니다!!😊

christy-hs-lee replied:
전역 변수란 함수의 외부에서 선언하여 어디에서나 접근할 수 있어 광범위하게 사용할 수 있는 변수입니다!

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,66 @@
+let initialNumbers = makeRandomNumbers()
 
+func makeRandomNumbers() -> [Int] {
+    var resultnumbers: Set<Int> = []
```
디스코드에서 이야기했듯이 [General Conventions](https://www.swift.org/documentation/api-design-guidelines/#general-conventions)를 읽어보시면 좋겠네요!
스네이크 케이스, 파스칼 케이스, 카멜 케이스가 무엇일까요?

christy-hs-lee replied:
스네이크 케이스는 _(언더바)를 사용하여 단어를 구분하는 표기법입니다! 
     ex) snake_case 처럼 사용합니다.

파스칼 케이스는 단어의 시작 부분을 모두 대문자로 시작되는 표기법입니다!
     ex) PascalCase 처럼 사용합니다.

카멜 케이스는 첫 글자는 소문자로 중간 글자들은 대문자로 작성하는 표기법입니다!
     ex) camelCase 처럼 사용합니다.

Gundy93 replied:
잘 조사해오셨으니 추가로!

[General Conventions](https://www.swift.org/documentation/api-design-guidelines/#general-conventions) 공식문서에서는 `UpperCamelCase`, `lowerCamelCase` 라고 표현하는데, 이 차이도 캐치하시면 네이밍할 때 기억에 남을 거라고 생각합니다 😃

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,66 @@
+func makeRandomNumbers() -> [Int] {
+    var resultnumbers: Set<Int> = []
+
+    while resultnumbers.count < 3 {
+        guard let randomElement = numbers.randomElement() else{
+            return []
+        }
+        resultnumbers.insert(randomElement)
+    }
```
`guard`문을 통해 안전하게 옵셔널을 처리하셨네요! 아주 잘하셨습니다 👍

`guard`와 `if`의 차이가 무엇이며, `guard`문의 `else` 안에는 `return` 밖에 사용할 수 없을까요?

juun97 replied:
guard 는 조건이 만족되지 않을 때 실행되고 if 는 조건이 만족될때 실행이 됩니다!
공식문서에서는 guard 를 빠른 종료를 위해 사용한다고 합니다! 그래서 else 안에 꼭 즉시 종료를 해주는 코드가 꼭 들어가있어야 해요
예시로는 return 도 있지만 continue, throw, break 등등이 있겠네요!
그리고 guard 문을 사용하게 되면 코드의 가독성이 높아져 사용을 권장한다고 하네요 😃😃

Gundy93 replied:
아주 잘 정리했군요! 좋습니다
다만 `guard`는 조건이 만족되지 않을 때 `else 블록이` 실행된다고 조금 덧붙이면 더 좋을 것 같아요!

잘 알아보셨으니 추가로 `return []`은 (설령 절대로 실행되지 않을 else문이더라도) 원하는 적절한 반환값이 아닐 것 같습니다
그렇다면 이 반복문에서 return말고 사용할 수 있는 것이 무엇이 있을까요?
(만약 이 질문이 잘 이해가 안된다면 return이 무엇인지에 대해서도 고민해볼까요?)

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,66 @@
+}
+
+func playGame() {
+    var temp = 9
```
`temp`라는 네이밍은 조금 모호한 것 같습니다
변수명이 길어지더라도 더욱 명확하게 작성하는 것이 좋을 것 같네요
남은 기회임을 더 명확히 알 수 있는 네이밍을 고민해볼까요?

그리고 제약사항에 따라 위치를 수정해볼 수도 있을 것 같아요 😃

juun97 replied:
검색을 해보니까 잔여의 뜻을 가진 단어가 residual 이더라구요! 뭔가 생소한 단어긴 하지만 residualChance 이런 식은 어때요 건디!?

Gundy93 replied:
`temp`보다 훨씬 명확한 의미인 것 같아서 좋습니다! 👍

다만 이제 생소하거나 널리 쓰이지 않는 단어라서 고민된다면 `remainingChances` 같은 네이밍도 괜찮을 것 같네요 😃

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,66 @@
+
+}
+
+func checkNumbers(numbers: [Int]) -> [Int] {
```
`checkNumbers(numbers: [Int])`라는 네이밍은 numbers라는 단어가 반복돼서 조금 어색한 것 같아요

`checkResult(of numbers: [Int])`와 같은 네이밍도 하나의 예시가 될 것 같습니다
더 좋은 네이밍도 고민해보자구요! 😃

juun97 replied:
건디말 듣고 나니 죄다 numbers 로 범벅되서 막 헷갈릴 수도 있을거 같아요! 조언 감사해요 건디😃

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,66 @@
+    var ball = 0
+
+
+    for number in numbers {
+        if initialNumbers.contains(number){
+            ball += 1
+        }
+    }
+    for index in 0...2{
+        if initialNumbers[index] == numbers[index]{
+            strike += 1
+            ball -= 1
+        }
+    }
```
반복문을 사용해 각 값에 대해 비교를 하셨군요 잘하셨습니다 👍

혹시 이 두 `for-in` 구문을 하나로 합칠 수 있을까요?

juun97 replied:
```swift
for index in 0...2{
       
       if initialNumbers.contains(numbers[index]){
           ball +=1
       }
       
       if initialNumbers[index] == numbers[index]{
           strike += 1
           ball -= 1
       }
   }
```
이렇게 바꿀 수 있을거 같네요!

Gundy93 replied:
반복문을 성공적으로 합쳤군요! 아주 잘하셨어요 👍

`if`, `else if` `else`에 대해서도 알아보시면 좋을 것 같습니다~
볼을 추가한 다음에 다시 빼지 않더라도 조건문을 작성할 수 있을 거에요 😃

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,66 @@
+    var temp = 9
+    var result : [Int] = []
+
+    while temp > 0 {
```
`Bool` 타입의 변수가 아닌 조건문을 통해서 `while`문을 사용하신거 정말 좋습니다! 👍

`Bool` 타입의 변수를 탈출조건으로 삼으면 위험한 이유가 무엇일까요?
(물론 조건문도 같은 위험이 있지만 가독성 측면에서는 더 명확하게 알아볼 것 같아요)

juun97 replied:
음 제가 생각했을땐 만약 Bool 타입의 변수가 끝까지 True를 유지하게 된다면 무한루프 상태에 빠져서 함수가 안끝날거 같아요!

Gundy93 reviewed (commented):
STEP 1 구현 잘 해주셨습니다 여러분 🙌
익숙하지 않은 Git과 commit, GitHub에 대해서 여러 시도를 하신 것 같네요 👍👍
디스코드로 여러분과 나눈 이야기에서는 잘 전달이 되었을지 모르겠지만 제 답변은 절대로 정답이 아닙니다!
코드에는 정답이 없죠! 이해가 되지 않는 피드백에 대해서는 같이 이야기를 나눠봐도 좋을 것 같아요
앞으로도 더 좋은 답을 찾는 과정이라고 생각하면서 코드를 작성해주시면 좋을 것 같아요 😃

---

## 고민했던 부분💡
고민을 하고, 해결을 하고자 하신 부분 칭찬합니다! 👍
구글링을 통해 '이런 것도 있구나~'하고 찾은 부분에 대해 공식 문서를 읽어보는 것도 좋을 것 같아요
하지만 공식 문서는 언제나 길고, 영어는 언제나 어렵죠...
아직은 Overview 정도만 읽어보는 것도 좋은 경험일 것 같습니다 😃
### 중복되지 않은 3개의 숫자
[Set](https://developer.apple.com/documentation/swift/set)는 요소의 중복을 허용하지 않는 Collection 타입이므로 이를 통해 중복되지 않은 값을 구하고자 한 부분이 아주 좋은 것 같아요!
```swift
while resultnumbers.count < 3 {
        guard let randomElement = numbers.randomElement() else{
            return []
        }
        resultnumbers.insert(randomElement)
    }
```
하지만 만약 `numbers.randomElement()`가 우연히도 같은 숫자만 나온다면 이 반복문은 끝나지 않을 것 같네요
이에 대해서도 한번 생각해보시고 더 좋은 방법이 없을지 잠깐정도는 고민해보시는 것도 좋을 것 같습니다
**너무 길지는 않게요**
### FlowChart 작성법
지금 작성해주신 흐름도는 아주 깔끔하고 좋은 것 같아요! 👍 **simple is the best**
흐름도는 전체적인 로직의 흐름을 이해하기 위한 그림이기 때문에 지금정도로만 작성해줘도 명쾌하게 한눈에 잘 들어오는 것 같아요<br>
![](https://i.imgur.com/2kFOcUQ.png)
여기서 `임 / 의의` 처럼 한 단어가 쪼개지지 않게끔 줄바꿈을 해주셨으면 더 좋지 않았을까 정도는 생각합니다
또 흐름도 시작하는 부분의 화살표만 색이 다른데 의도하신 건지 궁금합니다 🤔

---

## 조언이 필요한 부분 ❓
### 변수의 추가적인 생성
이 질문에 대해서는 디스코드에서 이야기를 했는데요, **가독성**이 떨어지지 않는 선에서 생각해보세요 😃
따로 변수 선언해주신 부분이 읽기 정말 좋았습니다 👍

---

추가적으로 이 시기에 공부하면 좋을만한 것을 생각해봤는데요!
프로젝트에 나와있는대로 [Naming](https://www.swift.org/documentation/api-design-guidelines/#naming), [Optionals, nil, Optional Binding](https://docs.swift.org/swift-book/LanguageGuide/TheBasics.html), [Collection Types](https://docs.swift.org/swift-book/LanguageGuide/CollectionTypes.html)를 잘 읽어보는 게 좋을 것 같습니다!
항상 프로젝트를 할 때 공부하기 좋은 녀석들이니까 잘 읽어보세요
또 읽어보시면 `Collection Types`에 소개되는 `Sets`의 경우 제가 보내드린 문서와 다른 문서일텐데요, 어떤 차이가 있는지 비교해보시는 것도 재밌을 것 같네요 😃

juun97 committed:
fix : 변수 Rename

reusltnumbers ->  reusltNumbers
temp -> leftChances
func checkNumbers() -> func checkResult()

juun97 committed:
refactor : makeRandomNumbers(). checkResult() 메서드 수정

makeRandomNumbers() 내부 guard문 수정
checkResult() 로직 내 반복문 수정
